### DISCUSSION TITLE
How to use a ramp down on MegaDrone or similar?

### DISCUSSION THREAD

**Discussion by YousefJayrwandi** [google/oboe#1196] [Q&A] [answered]
Hi @dturner and @philburk, 

Thank you for the quick replies to my question under the closed #326. I wanted to continue the thread here, as I had some questions about it, that may potentially benefit other inexperienced users. 
@dturner suggested that I used https://github.com/google/oboe/blob/master/src/flowgraph/RampLinear.h to continue outputting sound from the synth whenever the screen is released to avoid clicking sounds.
I made a few attempts to use it but without any successful result so was wondering if you could elaborate a little on where and how this RampLinear object could be used? And specifically as example in context of the MegaDrop sample (with an Oscillator that is modified to produce a sine wave instead of a square wave). Is the ramp applied somewhere in Oscillator.h renderAudio()
    mRamp.onProcess();
    audioData[i] = sin(mPhase)*mRamp.getTarget()
or is onProcess() used somewhere else, to modify the entire stream? 

@philburk suggested that I should leave the stream open for longer and mix more notes in, but I am not sure how to do that? The stream in MegaDrone is created in the engine and sound is running after a tap is detected and wave is set to on. 
is it some modification to the tappable source at Synth.h? 

void tap(bool isOn) override { for (auto &osc : mOscs) osc.setWaveOn(isOn);}
or again in renderAudio of the Oscillator.h, or else? 

I wanted to have a pure sine wave sound as well as one that is with mixed sine wave + a few harmonics, so the clicking sounds appear only when the harmonics are below 6 and the wave is relatively pure. I have no such problems when I have the main sine wave mixed with 16 harmonics for example or when using a square wave as is in MegaDrone by default. 

I am sorry if this is too basic, but I couldn't get it to work myself, so would really appreciate the help.
Thank you for your time.

**Comment by philburk** [comment, marked as answer]
You could use a RampLinear object. But that would involve switching over to using the flowgraph library and that is not well documented.

I would suggest making a smaller modification to your program.  Basically you just want to multiply your output by an amplitude variable. If amplitude=0.0 then the output will be silent.  If amplitude=1.0 then the output will be heard at full volume. Then ramp the amplitude up or down when you touch the screen.

    amplitude += amplitudeDelta;
    amplitude = std::min(1.0, std::max(0.0, amplitude)); // clip to range
    output = signal * amplitude;

When you tap down:

    amplitudeDelta = 0.001; // positive, so it will ramp up

When you lift your finger:

    amplitudeDelta = -0.001; // negative, so it will ramp down

> suggested that I should leave the stream open for longer and mix more notes in, but I am not sure how to do that? 

If you used MegaDrone then you are already doing that. Just start the stream once and leave it running while you change your sound synthesizer.

> the clicking sounds appear only when the harmonics are below 6 and the wave is relatively pure. 

Yes, clicking is more obvious when playing pure low frequency sine tones. Complex waves like square are all clicks so you don't notice the volume clicks.

**Reply by YousefJayrwandi** [reply]
Thank you very much for your reply. This was very helpful and it now works after slight modifications. 
The amplitudeDelta for me worked when it was set to +/-0.0001. For some reason, it was not ramping gradually enough and with +/-0.001 it was still clipping.  
Thank you for your time, my query has been resolved.
